private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -150,6 +188,92 @@ private static boolean isSchemeSupported(String repo) {
+
+    return download(mavenMetadataUrl, credentials).thenApply( optionalFileContents -> {
+      try {
+        if (optionalFileContents.isEmpty()) {
+          // no file so just upload a new one
+          // we must bootstrap
+          Metadata metadata = new Metadata();
+          metadata.setGroupId(coords.groupId);
+          metadata.setArtifactId(coords.artifactId);
+          metadata.setVersioning(new Versioning());
+          return metadata;
+        }
+          return new MetadataXpp3Reader().read(new StringReader(optionalFileContents.get()), false);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+
+    });
+  }
+
+  /**
+   * Upload the new maven-metadata.xml with the new version included in the version list &
+   * set the latest and release tags in the Metadata XML object.
+   * This function will first download the pre-existing metadata-xml and augment.
+   * If no maven-metadata.xml exists, a new one will be hydrated.
+   */
+  private static CompletableFuture<Void> uploadMavenMetadata(
+          String repo,
+          Credentials credentials,
+          Coordinates coords
+  ) {
+
+    String mavenMetadataUrl =
+            String.format(
+                    "%s/%s/%s/maven-metadata.xml",
+                    repo.replaceAll("/$", ""),
+                    coords.groupId.replace('.', '/'),
+                    coords.artifactId);
+    return downloadExistingMavenMetadata(repo, credentials, coords).thenCompose(
+            metadata -> {
+              try {
+
+                // There is a chance versioning is null; handle it by creating the empty object.
+                Versioning versioning = Optional.ofNullable(metadata.getVersioning()).orElse(new Versioning());
+                versioning.setLatest(coords.version);
+                versioning.setRelease(coords.version);
+                // This may be needed for SNAPSHOT support
+                String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+                versioning.setLastUpdated("20200731090423");
+                versioning.getVersions()
+                        .add(coords.version);
+                // Let's handle adding multiple versions many times by turning it back to a set
+                versioning.setVersions(
+                        versioning.getVersions()
+                                .stream()
+                                .distinct()
+                                .collect(Collectors.toList())
+                );
+                metadata.setVersioning(versioning);
+
+                Path newMavenMetadataXml = Files.createTempFile("maven-metadata", ".xml");
+                ByteArrayOutputStream os = new ByteArrayOutputStream();
+                new MetadataXpp3Writer().write(os, metadata);
+                Files.write(newMavenMetadataXml, os.toByteArray());
+                return upload(mavenMetadataUrl, credentials, newMavenMetadataXml);
```

Comment:
nit: upload could probably take InputStream but I didn't want to change too much of the code; 
Open to doing another pass after on this file if you are open to it.